test/functional/ui/incsubstitution_spec.lua

```diff
@@ -0,0 +1,1156 @@
+        two or three LZNES
+        ]])
+      end
+  end)
+
+  it('work with a failing mapping', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap <expr> x execute('bwipeout!')[-1].'x'")
+      source([[
+        func! DoIt()
+          bwipeout!
+        endfunc]])
+      --execute("cnoremap <expr> x DoIt()[-1].'x'")
+
+      feed(":%s/tw/tox<enter>")
+
+      -- error thrown b/c of the mapping
+      neq(nil, string.find(eval('v:errmsg'), '^E523:'))
+      -- the substitution after the error only works for ics=split/nosplit
+      -- which seems like the right thing to do in all cases, but we probably
+      -- don't want to change the default, so all in all this seems alright
+      if case == '' then
+        expect(default_text)
+      else
+        expect(string.gsub(default_text, "tw", "tox"))
+      end
+    end
+  end)
+
+  it('work when temporarily moving the cursor', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap <expr> x cursor(1, 1)[-1].'x'")
+
+      feed(":%s/tw/tox/g<enter>")
+      expect(string.gsub(default_text, "tw", "tox"))
+    end
+  end)
+
+  it('work when a mapping disables incsub', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      execute("cnoremap <expr> x execute('set incsubstitute=')[-1]")
+
+      feed(":%s/tw/toxa/g<enter>")
+      expect(string.gsub(default_text, "tw", "toa"))
+    end
+  end)
+
+  it('work with a complex mapping', function()
+    for _, case in pairs(cases) do
+      refresh(case)
+      source([[cnoremap x <C-\>eextend(g:, {'fo': getcmdline()})
+      \.fo<CR><C-c>:new<CR>:bw!<CR>:<C-r>=remove(g:, 'fo')<CR>x]])
+
+      feed(":%s/tw/tox")
+      feed("/<enter>")
+      expect(string.gsub(default_text, "tw", "tox"))
+    end
+  end)
+
+end)
+
+-- *** AUTOCMDS
+-- Right now, none fire when opening or closing the incsub buffer
+--
+-- If that behavior is changed, add tests (suggestions by ZyX-l):
+-- * If any are run when window is opened, then tests which check what if these
+--   autocommands close the window.
+-- * If any are run when window is opened, then
+--   tests which check what if these autocommands wipe out buffer without closing
+--   the window.
+-- * If any are run when window is opened, then tests which check
+--   what if these autocommands switch buffer without closing the window or
+--   wiping buffer out.
+-- * If any are run when window is opened, then tests which
+--   check what if these autocommands wipe out buffer where substitution is
+--   being made.
+-- * If any are run when window is opened, then tests which check what
+--   if these autocommands close window where substitution is being made (but
+--   not window with temporary view)
+
+describe('Incsubstitute: autocommands', function()
+  before_each(clear)
+
+  -- never, always, openOnly, closeOnly
+  local eventsExpected = {
+    ["BufAdd"] = "never",
+    ["BufDelete"] = "openOnly",
```

Comment:
Can you also add buffer numbers to the test (`expand('<abuf>')`)? It is rather strange to see `BufDelete='openOnly'` and it would be good to check for which buffer(s) it fiers. Also you do not need `["` and `"]`.